Network/Socket/Types.hsc

```diff
@@ -112,11 +117,53 @@ invalidateSocket (Socket ref _) errorAction normalAction = do
 
 -----------------------------------------------------------------------------
 
--- | Close the socket. Sending data to or receiving data from closed socket
+-- | Close the socket. This function does not throw exceptions even if
+--   the underlining system call returns errors.
+--
+--   Sending data to or receiving data from closed socket
 --   may lead to undefined behaviour.
+--
+--   This function is not thread-safe. Consider the following senario.
+--
+--   1) Thread A acquires a 'Fd' from 'Socket' by 'fdSocket'.
+--
+--   2) Thread B close the 'Socket'.
+--
+--   3) Thread C opens a new 'Socket'. Unfortunately it gets the same 'Fd'
+--      number which thread A is holding.
+--
+--   In this senario, thread A continues its work on the inappropriate 'Fd'.
+
 close :: Socket -> IO ()
 close s = invalidateSocket s (\_ -> return ()) $ \oldfd -> do
-    closeFdWith (void . c_close . fromIntegral) (fromIntegral oldfd)
+    -- closeFdWith avoids the deadlock of IO manager.
+    closeFdWith closeFd (toFd oldfd) `E.catch` ignore
```

Comment:
Does this also catch asynchronous exceptions?  It would be better if we had a low-level close that returned an "int", and throw an exception in `close'` and not throw in `close`, rather than try to catch exceptions here.

Reply:
That is use `c_close` here and ignore the return value!

Reply:
> That is use c_close here and ignore the return value!

Ah. You are right. `c_close` does not throw exceptions!

Reply:
> Does this also catch asynchronous exceptions?

What kind of scenario do you think?

Reply:
Things like timeouts and thread cancellation are asynchronous timeouts and must not be caught here.  Michael Snoyman has a Control.Exception.Safe module that can catch just synchronous exceptions, but there's no need for that here...

http://hackage.haskell.org/package/safe-exceptions-0.1.7.0

Reply:
`ignore` was removed.
We should consider asynchronous exceptions yet, if necessary.

Reply:
OK. I understand what you want to explain. I hope that you like the current code.